### DISCUSSION TITLE
Proposal: a `batch` variant for wrapping callback functions

### DISCUSSION THREAD

**Discussion by Mox93** [solidjs/solid#1736] [Ideas]
> Updated the description to make it less confusing

It would be nice to have a `batchCallback` version of the `batch` function that wraps a callback directly (the same way React provides `useCallback` even though you could achieve the same thing with `useMemo`). The reason behind that is it seems that `batch` is mostly used with callback functions similar to how it's explained in the [tutorials](https://www.solidjs.com/tutorial/reactivity_batch). So my proposal is to add something like this:
```typescript
function batchCallback<T, A extends T[], R, F extends (...args: A) => R>(
  callback: F
) {
  return (...args: A) => batch(() => callback(...args));
}
```

**Comment by snnsnn** [comment]
> The reason behind that is it seems that batch is mostly used with callback functions

`batch` is used to update multiple signals at once so that you don't cause multiple re-renders. useCallback has no place in Solid because components are rendered once, when they are loaded the first time.

**Comment by Mox93** [comment]
@snnsnn, I'm not comparing `batch` to `useCallback` and I'm aware Solid doesn't need a `useCallback`. What I'm suggesting is to have a version of `batch` that takes a function and returns a new function with the same signature it took as input.
So instead of doing this:
```javascript
const updateNames = () => {
    console.log("Button Clicked");
    batch(() => {
        setFirstName(firstName() + "n");
        setLastName(lastName() + "!");
    })
}
```
we'd do something along these lines:
 ```javascript
const updateNames = batchCallback(() => {
    console.log("Button Clicked");
    setFirstName(firstName() + "n");
    setLastName(lastName() + "!");
})
```
It's just a small change to write less code.

**Reply by snnsnn** [reply]
It seems I did misunderstood you however batch is a sync function, so you can use it directly:
```ts
const updateNames = batch(() => {
   console.log("Button Clicked");
   setFirstName(firstName() + "n");
   setLastName(lastName() + "!");
})
```
I don't think it is a good idea to introduce an additional layer of indirection.

**Reply by Mox93** [reply]
> so you can use it directly:
> ```javascript
> const updateNames = batch(() => {
>    console.log("Button Clicked");
>    setFirstName(firstName() + "n");
>    setLastName(lastName() + "!");
> })
> ```

And what will the value of `updateNames` be in that case? Cause I'm sure if you test it out, it won't work.
Here you go https://playground.solidjs.com/anonymous/db414fdf-8e81-4c4b-bf91-ad2b5179cb63 (this is how your solution looks).

**Reply by snnsnn** [reply]
Yes, it should be:

```
 const updateNames = () => batch(() => {
    console.log("Button Clicked");
    setFirstName(firstName() + "n");
    setLastName(lastName() + "!");
  });
```

But your solution saves few characters at best in exchange of reduced performance because we have destructuring and spread by default, both are quite slow operations.
